pymongo/asynchronous/client_bulk.py

```diff
@@ -0,0 +1,763 @@
+                        # Synthesize the full bulk result without modifying the
+                        # current one because this write operation may be retried.
+                        full = copy.deepcopy(full_result)
+                        _merge_command(self.ops, self.idx_offset, full, result)
+                        _throw_client_bulk_write_exception(full, self.verbose_results)
+
+                # Process the server reply as a command cursor.
+                await self._process_results_cursor(full_result, result, conn, session)
+
+                # Merge this batch's results with the full results.
+                _merge_command(self.ops, self.idx_offset, full_result, result)
+
+                # We're no longer in a retry once a command succeeds.
+                self.retrying = False
+                self.started_retryable_write = False
+
+            else:
+                to_send_ops, _ = await self._execute_batch_unack(bwc, cmd, ops)  # type: ignore[arg-type]
+
+            self.idx_offset += len(to_send_ops)
+
+            # We halt execution if we hit a top-level error,
+            # or an individual error in an ordered bulk write.
+            if full_result["error"] or (self.ordered and full_result["writeErrors"]):
+                break
+
+    async def execute_command(
+        self,
+        session: Optional[AsyncClientSession],
+        operation: str,
+    ) -> MutableMapping[str, Any]:
+        """Execute commands with w=1 WriteConcern."""
+        full_result: MutableMapping[str, Any] = {
+            "anySuccessful": False,
+            "error": None,
+            "writeErrors": [],
+            "writeConcernErrors": [],
+            "nInserted": 0,
+            "nUpserted": 0,
+            "nMatched": 0,
+            "nModified": 0,
+            "nDeleted": 0,
+            "insertResults": {},
+            "updateResults": {},
+            "deleteResults": {},
+        }
+        op_id = _randint()
+
+        async def retryable_bulk(
+            session: Optional[AsyncClientSession],
+            conn: AsyncConnection,
+            retryable: bool,
+        ) -> None:
+            if conn.max_wire_version < 25:
+                raise InvalidOperation(
+                    "MongoClient.bulk_write requires MongoDB server version 8.0+."
+                )
+            await self._execute_command(
+                self.write_concern,
+                session,
+                conn,
+                op_id,
+                retryable,
+                full_result,
+            )
+
+        await self.client._retryable_write(
+            self.is_retryable,
+            retryable_bulk,
+            session,
+            operation,
+            bulk=self,  # type: ignore[arg-type]
+            operation_id=op_id,
+        )
+
+        if full_result["error"] or full_result["writeErrors"] or full_result["writeConcernErrors"]:
+            _throw_client_bulk_write_exception(full_result, self.verbose_results)
```

Comment:
What error gets raise when a batch fails with a top-level command error or a connection error (eg AutoReconnect)? The coll.bulk_write api suffers from a problem where any batch that fails with such an error (after retries) will abandon the bulk write exception and raise AutoReconnect directly, thus loosing information about previous batches that could have succeeded. 

If the new bulk_write suffers from the same problem we should open a follow up ticket to improve the behavior.

Reply:
I opened https://jira.mongodb.org/browse/PYTHON-4598, so I have something to point to when skipping the relevant tests in [PYTHON-4242](https://jira.mongodb.org/browse/PYTHON-4242).

Reply:
A `ClientBulkWriteException` gets raised for any top-level command/connection errors, and the `error` field of the exception will contain info about the error. So if it's a connection error like AutoReconnect, the `error` field will contain the error itself. If it's a top-level command error, the `error` field will contain the raw server response. Any results from previous batches get stored in the `partial_result` field of the `ClientBulkWriteException`.